### PR TITLE
WIP: Metadata system take 2

### PR DESCRIPTION
**Pull request by shashi**
Now always keyed by a context type.

### CODE REVIEW

File: src/types.jl [hunk 1]
```diff
@@ -56,6 +56,55 @@ symtype(x) = typeof(x)
...
+    end
+    Base.ImmutableDict{DataType, Any}(d, ctx, val)
+end
+
+function setmetadata(s::Symbolic, ctx::DataType, val)
+    if s.metadata isa Ref
+        @set s.metadata = Ref{Any}(assoc(s.metadata[], ctx, val))
+    else
+        # fresh Dict
+        @set s.metadata = Ref{Any}(Base.ImmutableDict{DataType, Any}(ctx, val))
```
Comment (shashi) [inline src/types.jl:104]: @YingboMa I think this Ref is not necessary here, it will be `Union{Nothing, ImmutableDict{DataType, Any}}`.

Comment (shashi) [reply src/types.jl:104]: Just as good. In fact we can even use `ImmutableDict{DataType, Any}()` as the default "empty" thing, which would mean no type parameter. But that empty struct adds 24 bytes to Sym... Which maybe just fine!

Comment (YingboMa) [reply src/types.jl]: Have you benchmarked it?

Comment (shashi) [reply src/types.jl]: No I didn't benchmark the empty ImmutableDict, right now it is still a parameter which can be Nothing or ImmutableDict.

Comment (shashi) [reply src/types.jl]: This is faster than having a Ref of ImmutableDict.

**Pull request merged**